lib/sgr.js

```diff
@@ -0,0 +1,115 @@
+'use strict';
+/* CSI - control sequence introducer */
+/* SGR - set graphic rendition */
+
+var assign   = require('es5-ext/object/assign')
+  , includes = require('es5-ext/string/#/contains');
+
+var CSI = '\x1b[';
+
+var sgr = function (code) {
+	return CSI + code + 'm';
+};
+
+sgr.CSI = CSI;
+
+var mods = assign({
+	// Style
+	bold:      { _bold: [1, 22] },
+	italic:    { _italic: [3, 23] },
+	underline: { _underline: [4, 24] },
+	blink:     { _blink: [5, 25] },
+	inverse:   { _inverse: [7, 27] },
+	strike:    { _strike: [9, 29] }
+
+	// Color
+}, ['black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white']
+	.reduce(function (obj, color, index) {
+		// foreground
+		obj[color] = { _fg: [30 + index, 39] };
+		obj[color + 'Bright'] = { _fg: [90 + index, 39] };
+
+		// background
+		obj['bg' + color[0].toUpperCase() + color.slice(1)] = { _bg: [40 + index, 49] };
+		obj['bg' + color[0].toUpperCase() + color.slice(1) + 'Bright'] = { _bg: [100 + index, 49] };
+
+		return obj;
+	}, {}));
+
+sgr.mods = mods;
+
+(function () {
+	var forEach = require('es5-ext/object/for-each')
+	  , onlyKey = require('es5-ext/object/first-key')
+	  , uniq = require('es5-ext/array/#/uniq.js');
+
+	sgr.toModPair = function (mod) {
+		return mod[onlyKey(mod)];
+	};
+
+	var openers = {}, closers = {};
+
+	forEach(mods, function (mod) {
+		var modPair = sgr.toModPair(mod);
+
+		openers[modPair[0]] = modPair;
+		closers[modPair[1]] = modPair;
+	});
+
+	sgr.openers = openers;
+	sgr.closers = closers;
+
+	sgr.openStyle = function (mods, code) {
+		mods.push(sgr.openers[code]);
+	};
+
+	sgr.closeStyle = function (mods, code) {
+		mods.slice().forEach(function (modPair, index) {
+			if (modPair[1] === code) {
+				mods.splice(index, 1);
+			}
+		});
```

Comment:
Instead of iterating on copy, we can iterate in reverse direction:
```javascript
var forEachRight = require('es5-ext/array/#/for-each-right');
forEachRight.call(mods, function (modPair, index) {
  if (modPair[1] === code) mods.splice(index, 1);
});
```

or if we do not need to have `mods.length` accurate, we can just do:
```javascript
mods.forEach(function (modPair, index) {
  if (modPair[1] === code) delete mods[index];
});
```